src/parse.ts

```diff
@@ -1,119 +1,120 @@
-		depth = indent.length + 1;
-	}
-	return root;
+export function parse(input: any) {
+	return new Parser(defaultSchema).parse(input);
 }
 
-function makeParseLine(schema: Schema): (input: string) => Line {
-	const TOKEN_COUNT = INLINE_TOKEN._size + schema.customTokens.length;
-	const INLINE = generateInlineRegex(schema);
+export class Parser {
+	private lineRegex = /(?:\r\n|\n|\r|^)(\t*)(?:#([^ \[\r\n]+)(?: |$))?(.*)/gm;
+	private lexLine: Lexer;
+
+	constructor(private schema: Schema) {
+		this.lexLine = compileLexer(
+			([
+				{
+					type: InlineToken.Escaped,
+					regex: /\\(.)/
+				},
+				{
+					type: InlineToken.OpenTag,
+					regex: /#([^ \[]+)(\[)?/
+				},
+				{
+					type: InlineToken.ArgSeparator,
+					regex: /]\[/
+				},
+				{
+					type: InlineToken.CloseTag,
+					regex: /]/
+				}
+			] as TokensDef[]).concat(
+				schema.customTokens.map(_ => ({
+					type: _.tag,
+					regex: _.regex
+				}))
+			),
+			InlineToken.Text
+		);
+	}
 
-	return (input: string): Line => {
+	parse(input: any): Block {
+		let depth = 0;
+		const root: Block = { tag: "root", children: [], head: [] };
+		const stack = [root];
+		let current = root;
+		for (const [_, indent, tagName, lineContent] of input.matchAll(this.lineRegex)) {
+			if (!tagName && !lineContent) continue;
+			for (let i = 0; i < depth - indent.length; ++i) stack.pop();
+			current = last(stack);
+			const tag = tagName || this.schema.getDefault(current.tag);
+			const head = this.parseLine(lineContent);
+			const block = { tag, head, children: [] };
+			current.children.push(block);
+			stack.push(block);
+			depth = indent.length + 1;
+		}
+		return root;
+	}
+
+	parseLine(input: string): Line {
+		if (!input) return [];
 		const root: Line = [];
 		const inlineStack: Inline[] = [];
 		let current: Line = root;
-		let inInline = false;
-		const tokens = input.split(INLINE);
-		for (let i = 0; i < tokens.length; ++i) {
-			const token = tokens[i];
-			if (!token) continue;
-			switch (i % TOKEN_COUNT) {
-				case INLINE_TOKEN.text:
-				case INLINE_TOKEN.escaped:
-					current.push(token);
+		for (const token of this.lexLine(input)) {
+			switch (token.type) {
+				case InlineToken.Text:
+					current.push(token.value);
 					break;
-				case INLINE_TOKEN.tagName:
-					const inlineAst = { tag: token, arguments: [], closed: false };
-					inlineStack.push(inlineAst);
-					current.push(inlineAst);
-					inInline = true;
+				case InlineToken.Escaped:
+					current.push(token.matches[0]!);
 					break;
-				case INLINE_TOKEN.openArg:
-				case INLINE_TOKEN.nextArg:
-					if (inInline) {
-						current = appendArg(last(inlineStack));
+				case InlineToken.OpenTag:
+					const inlineAst = { tag: token.matches[0]!, arguments: [], closed: false };
+					current.push(inlineAst);
+					if (!token.matches[1]) {
+						break;
 					} else {
-						current.push(token);
+						inlineStack.push(inlineAst);
+					}
+				case InlineToken.ArgSeparator:
+					if (inlineStack.length) {
```

Comment:
More readable if written as `inlineStack.length > 0`. Same applies for the next case